### PR TITLE
[lldb] Get lldb with swift support building on stable/20230725 (Darwin)

### PR DESCRIPTION
**Pull request by bulbazord**
I've done the necessary work to get this building on Darwin. This PR has 2 kinds of patches:
1) Patches I wrote to adjust the swift-specific code to adapt to changes that happened upstream
2) Cherry-picks that were not applied to internal/next before the rebranch

To handle (2), I compared the source code of stable/20230725 and stable/20221013 and figured out which patches were primarily responsible for the diffs. I cherry-picked the relevant ones to get this branch building. I explicitly did NOT go through the history to figure out which ones should be applied (for correctness, for new features, etc). That will have to be done separately.

### CODE REVIEW

**Review by medismailben — APPROVED** [review]
LGTM!

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h [hunk 1]
```diff
@@ -59,6 +59,12 @@ class DWARFASTParserSwift : public DWARFASTParser {
   void EnsureAllDIEsInDeclContextHaveBeenParsed(
       lldb_private::CompilerDeclContext decl_context) override {}
 
+  // FIXME: What should this do?
+  lldb_private::ConstString
+  GetDIEClassTemplateParams(const DWARFDIE &die) override {
+    return lldb_private::ConstString();
```
Comment (compnerd) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h:67]: Hmm, this seems like a silent time bomb.

**Review by compnerd — APPROVED** [review]

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 2]
```diff
@@ -4163,7 +4163,6 @@ static SwiftASTContext::TypeOrDecl DeclToTypeOrDecl(swift::ASTContext *ast,
     case swift::DeclKind::Module:
     case swift::DeclKind::Missing:
     case swift::DeclKind::MissingMember:
-    case swift::DeclKind::MacroExpansion:
```
Comment (etcwilde) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:4166]: Where did the MacroExpansion decl kind go?
It's in the list of DeclKinds as far as I can tell. 
https://github.com/apple/swift/blob/f758144b55ec7f6846f3a2653935b7c3934d2ac3/include/swift/AST/Decl.h#L127
https://github.com/apple/swift/blob/f758144b55ec7f6846f3a2653935b7c3934d2ac3/include/swift/AST/DeclNodes.def#L195

Comment (bulbazord) [reply lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:4166]: This was a duplicate. Github conveniently left out the context, scroll up 4 lines and you'll see it's already listed.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h (hunk 1 above)
Comment (bulbazord) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h:67]: Yeah, likely. Not sure what this _should_ return. I plan on leaving this as-is for now and letting @adrian-prantl figure out what to do with it.

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp (hunk 2 above)
Comment (etcwilde) [reply lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:4166]: Ah, there it is. Thanks.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h (hunk 1 above)
Comment (etcwilde) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h:67]: Is this expected to get hit? Would it make sense to either have an `assert(false && "I don't know what I'm doing here");` or an `llvm_unreachable`? Seems like a nice crash somewhere would be helpful for debugging purposes if it ever does get hit.

**Review by etcwilde — APPROVED** [review]
I think other than the possible time bomb of pain, LGTM as far as I can tell.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h (hunk 1 above)
Comment (compnerd) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h:67]: Yeah, a more obvious NYI would be nice (e.g. `assert(false);`).

Comment (bulbazord) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.h:67]: Added an assertion

**Comment by bulbazord** [comment]
Going to land this now. We can change things after we get the branch building.

**Comment by bulbazord** [comment]
Backporting to `next`: https://github.com/apple/llvm-project/pull/7199

**Pull request merged**